addons/mail/static/src/messaging/entity/entity/field.js

```diff
@@ -0,0 +1,796 @@
+odoo.define('mail.messaging.entity.Field', function (require) {
+'use strict';
+
+/**
+ * Class whose instances represent field on a entity class.
+ * These field definitions are generated from declared fields in static prop
+ * `fields` on the entity classes.
+ */
+class Field {
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    constructor({
+        autocreate = false,
+        compute,
+        default: def,
+        dependencies = [],
+        dependents = [],
+        fieldName,
+        fieldType,
+        hashes: extraHashes = [],
+        inverse,
+        isCausal = false,
+        related,
+        relationType,
+        to,
+    } = {}) {
+        const id = _.uniqueId('field_');
+        /**
+         * This prop only makes sense for fields of type "relation". If set,
+         * it automatically creates a new entity for this field on creation of
+         * entity, and auto-links with this entity.
+         */
+        this.autocreate = autocreate;
+        /**
+         * If set, this field acts as a computed field, and this prop
+         * contains the name of the instance method that computes the value
+         * for this field. This compute method is called on creation of entity
+         * and whenever some of its dependencies change. @see dependencies
+         */
+        this.compute = compute;
+        /**
+         * Default value for this field. Used on creation of this field, to
+         * set a value by default.
+         */
+        this.default = def;
+        /**
+         * List of field on current entity that this field depends on for its
+         * `compute` method. Useful to determine whether this field should be
+         * registered for recomputation when some entity fields have changed.
+         * This list must be declared in entity definition, or compute method
+         * is only computed once.
+         */
+        this.dependencies = dependencies;
+        /**
+         * List of fields that are dependent of this field. They should never
+         * be declared, and are automatically generated while normalizing
+         * fields. This is populated by compute `dependencies` and `related`.
+         */
+        this.dependents = dependents;
+        /**
+         * Name of the field in the definition of fields on entity.
+         */
+        this.fieldName = fieldName;
+        /**
+         * Type of this field. 2 types of fields are currently supported:
+         *
+         *   1. 'attribute': fields that store primitive values like integers,
+         *                   booleans, strings, objects, array, etc.
+         *
+         *   2. 'relation': fields that relate to some other entities.
+         *
+         * Rule of thumb is to always use 'relation' when possible.
```

Comment:
just removed comment, because badly worded and it doesn't fit here. Initial intent was to hint of not defining an 'attribute' whose value contain entities, but that's common sense and I think the raises exception on attempting to write on a field with value that contains entity is clear enough.